diff --git a/lisp/net/tramp-gvfs.el b/lisp/net/tramp-gvfs.el
index a4a7bacd8a..b4263df150 100644
--- a/lisp/net/tramp-gvfs.el
+++ b/lisp/net/tramp-gvfs.el
@@ -123,6 +123,7 @@ tramp-gvfs-enabled
 	 (tramp-compat-funcall 'dbus-get-unique-name :system)
 	 (tramp-compat-funcall 'dbus-get-unique-name :session)
 	 (or (tramp-process-running-p "gvfs-fuse-daemon")
+             (tramp-process-running-p ".gvfsd-fuse-wrapped")
 	     (tramp-process-running-p "gvfsd-fuse"))))
   "Non-nil when GVFS is available.")
 
