diff --git a/lisp/emacs-lisp/bytecomp.el b/lisp/emacs-lisp/bytecomp.el
index 6c5051d70c4..be079346911 100644
--- a/lisp/emacs-lisp/bytecomp.el
+++ b/lisp/emacs-lisp/bytecomp.el
@@ -3570,8 +3570,6 @@ byte-compile-form
                 (setq form (cons 'progn (cdr form)))
                 (setq handler #'byte-compile-progn))
                ((and (or sef (function-get (car form) 'important-return-value))
-                     ;; Don't warn for arguments to `ignore'.
-                     (not (eq byte-compile--for-effect 'for-effect-no-warn))
                      (byte-compile-warning-enabled-p
                       'ignored-return-value (car form)))
                 (byte-compile-warn-x
-- 
2.42.0

